### ISSUE TITLE
Use a build system tool/monorepo manager for better performance and configurability

### ISSUE THREAD

**Issue by o-az** [huggingface/huggingface.js#48]
The best option IMO is https://turbo.build/repo/docs

> Turborepo is an intelligent build system optimized for JavaScript and TypeScript codebases.

> Turborepo leverages advanced build system techniques to speed up development, both on your local machine and your CI/CD.

Having used it myself, it's a fantastic tool that radically speeds up build/lint/test tasks and makes working with monorepos a breeze. It integrates with `pnpm` workspaces quite nicely.

It can also easily configure tasks that are dependent on other tasks (eg, only build package 'A' once package 'B' is done building).

I'm happy to set it up and open a PR to have you take a look.

**Comment by TimMikeladze** [comment]
Having any form for additional tooling or management over the packages, publishing, etc would be great.

**Comment by o-az** [comment]
Agreed. I'll integrate `turborepo` then 👍🏽

**Comment by coyotte508** [comment]
> It can also easily configure tasks that are dependent on other tasks (eg, only build package 'A' once package 'B' is done building).

For now we don't have interdependencies between repos

> Having used it myself, it's a fantastic tool that radically speeds up build/lint/test tasks

From what I see in the docs, it only speeds up building, specifically the bundling? The libraries are lightweight for now, with a single dependency.

> and makes working with monorepos a breeze. It integrates with pnpm workspaces quite nicely.

We can already lint / test the whole codebase with `pnpm -r format` or `pnpm -r test` in the root folder.

---

I think it's a bit too early to add this tooling, for now pnpm does the job fine, and it seems `changesets` is going to be added to manage releases :)

**Comment by o-az** [comment]
@coyotte508 From https://github.com/huggingface/huggingface.js/blob/main/.github/workflows/hub-publish.yml it looks like `scripts` in `docs-internal` are ran first then `pnpm publish` hub is executed. Can those scripts run in any order or does the `internal-docs` `prepublish-hub` need to run first?

Turborepo's filters can help with that  https://turbo.build/repo/docs/core-concepts/monorepos/filtering -- (this is also possible with `pnpm --filter`).

Here's an example of turborepo and changesets integration: https://github.com/vercel/turbo/tree/main/examples/with-changesets. Either way turborepo can be added at a later point like you said. 

Side point: should configurations like eslint, tsconfig, and prettier be consolidated into one config file each? For example, a central `.eslintrc.cjs` and letting individual packages extend it. For packages that have specific rules they can easily override the rule from the base config file.

The benefit of this is consistency among the packages and less boilerplate. Here's how it would look like for eslint for example:

`packages/hub/package.json`
```diff
"devDependencies": {
	"eslint": "^8.33.0",
-	"@typescript-eslint/eslint-plugin": "^5.51.0",
-	"@typescript-eslint/parser": "^5.51.0",
-	"eslint-config-prettier": "^8.6.0",
-	"eslint-plugin-prettier": "^4.2.1",
+   "eslint-config-base": "workspace:*"
}
```
`packages/hub/.eslintrc.cjs`
```diff
module.exports = {
-	extends: ["eslint:recommended", "plugin:@typescript-eslint/recommended", "prettier"],
-	parser:  "@typescript-eslint/parser",
-	plugins: ["@typescript-eslint", "prettier"],
	root:    true,
-	env:     {
-		browser: true,
-		amd:     true,
-		node:    true,
-	},
-	rules: {
-		"no-constant-condition":                             "off",
-		"@typescript-eslint/no-empty-function":              "off",
-		"@typescript-eslint/explicit-module-boundary-types": "error",
-		"@typescript-eslint/consistent-type-imports":        "error",
-		"@typescript-eslint/key-spacing":                    ["error", { align: "value" }],
-		// For doc purposes, prefer interfaces
-		"@typescript-eslint/consistent-type-definitions":    ["error", "interface"],
-	},
+  "extends": ["base"]
};
```
`packages/eslint-config-base/index.cjs`
```js
module.exports = {
	extends: ["eslint:recommended", "plugin:@typescript-eslint/recommended", "prettier"],
	parser:  "@typescript-eslint/parser",
	plugins: ["@typescript-eslint", "prettier"],
	// ...etc
};
```

I can create an issue to discuss this in more details.

**Comment by coyotte508** [comment]
> Side point: should configurations like eslint, tsconfig, and prettier be consolidated into one config file each? For example, a central `.eslintrc.cjs` and letting individual packages extend it. For packages that have specific rules they can easily override the rule from the base config file.

See https://github.com/huggingface/huggingface.js/issues/1 :smiling_face_with_tear:

> @coyotte508 From https://github.com/huggingface/huggingface.js/blob/main/.github/workflows/hub-publish.yml it looks like `scripts` in `docs-internal` are ran first then `pnpm publish` hub is executed. Can those scripts run in any order or does the `internal-docs` `prepublish-hub` need to run first?

The `publish` is last to make sure there's no error generating the commit that is pushed later:

- generate commit (with `yarn version`, generating docs)
- publish to npm
- git push

But the contents of the docs do not appear in the published package. If we go ahead with `changesets` as described in #6 , the CI wouldn't push directly to `main` anyway - it'd be done by merging a PR. So it would become completely unrelated, the docs would continually be generated as the PR is updated, and the publish would happen in a separate CI when the PR is merged.

**Comment by pngwn** [comment]
This feels like it would be solving a problem we don't have, the CI is fast enough and more tooling always adds more complexity. Pnpm alone is fine for simple repos like this. 

As an aside we have had a lot of issues with turbo on the sveltekit repo because of remote caching and it not correctly recognising when to bust the cache. 

I wouldn't over complicate the tooling this early.

**Comment by pngwn** [comment]
Fwiw pnpm already does some topological sorting when running recursive commands, it isn't perfect but only really trips up when there are cycles which isn't that common. It also logs a warning in those cases.

**Comment by pngwn** [comment]
> Side point: should configurations like eslint, tsconfig, and prettier be consolidated into one config file each? For example, a central .eslintrc.cjs and letting individual packages extend it. For packages that have specific rules they can easily override the rule from the base config file.

In general you want to avoid this as much as possible, keeping things consistent is important. You don't want to get into a situation where different packages have different code standards. It is sometime necessary is the packages are of a different type, like say libraries and apps but other than that it should almost never happen. 

If a lint rule needs tweaking for one packages then it should typically be done in the main config, not just for a single package. fwiw i've stopped linting most of my projects now, it doesn't have as much value today as it used to although I'm not against it in principle.

**Comment by o-az** [comment]
> > Side point: should configurations like eslint, tsconfig, and prettier be consolidated into one config file each? For example, a central .eslintrc.cjs and letting individual packages extend it. For packages that have specific rules they can easily override the rule from the base config file.
> 
> In general you want to avoid this as much as possible, keeping things consistent is important. You don't want to get into a situation where different packages have different code standards. It is sometime necessary is the packages are of a different type, like say libraries and apps but other than that it should almost never happen.
> 
> If a lint rule needs tweaking for one packages then it should typically be done in the main config, not just for a single package. fwiw i've stopped linting most of my projects now, it doesn't have as much value today as it used to although I'm not against it in principle.

What I was thinking of is having a single eslint config file for consistency and if, for example, say you have one package that has some graphql and you want to install a graphql eslint plugin, then that would be installed in that specific package and not in the central config, assuming only that package has graphql code and no other package does. I totally agree on the consistency part - I wouldn't want different packages having different configs for the same rule.

**Comment by pngwn** [comment]
Not really sure what you would gain by only installing that plugin for the one package except creating a greater maintenance burden because if some other package started using graphql later you'd _also_ want it to use the same linting plugin but they could easily skip that step and it would be hard to catch in review because you don't see code that doesn't get added. 

If you just add everything to root someone would have to remove it to avoid those standards, which would easily be caught in review.

Regardless, I don't think is really a problem we have right now so it isn't something we should solve for.

**Comment by coyotte508** [comment]
> Regardless, I don't think is really a problem we have right now so it isn't something we should solve for.

Yes! Right now the repo is more of a collection of libraries - just two libraries not depending on each other.
